### PR TITLE
Fix "Please adopt containerBackgroundAPI" for iOS 17 Widgets

### PR DESCRIPTION
**Pull request by staskus**
Partially addresses #21672

The goal of this PR is to conform to the new iOS 17 Widget APIs to avoid "Please adopt containerBackgroundAPI" message:
- Implement an empty `containerBackground` for conformance. This API does not support the usage of `AccessoryWidgetBackground`, therefore I decided to remove the background altogether to fit the current design guidelines
- Update UI of HomeScreenWidget to support removable background
- Temporarily disabled iOS 17 margins to support StandBy mode

I want to make even better iOS 17 support in the short-term future. Plus, we probably should make naming changes, given HomeScreen widgets work on iPadOS Lock Screen as well.

## To test:

- Lock Screen and Home screen widgets should work as before
- Home Screen widgets and Lock Screen Widgets should now be available on iPadOS Lock Screen vertical and horizontal versions

image

## Regression Notes
1. Potential unintended areas of impact

Should be none, Widgets should work and appear as before

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by staskus** [comment]
@guarani 

This is not a high priority but I wanted to at least conform to these APIs so we wouldn't need to make quick beta fixes when CI is updated to Xcode 15.

**Comment by staskus** [comment]
@guarani review is not necessary right now.

CI throws an issue:

> value of type 'some WidgetConfiguration' has no member 'contentMarginsDisabled'

It means that we cannot make these changes until Xcode is updated on CI, so we cannot merge it right now.

File: WordPress/JetpackStatsWidgets/Widgets/HomeWidgetAllTime.swift
Comment (staskus) [inline WordPress/JetpackStatsWidgets/Widgets/HomeWidgetAllTime.swift]: It doesn't work

**Review by guarani — COMMENTED** [review]
To summarize what we discussed in Slack, p1697208501951699/1697187055.060939-slack-C441E3YTS, once`containerBackground` was implemented, Lock Screen widgets no longer showed the `Please adopt containerBackground API`, but Home Screen widgets still do, despite implementing `containerBackgroundRemovable` and passing in `false`.

**Comment by staskus** [comment]
## Home Widgets (System Small)

I added to support background removal APIs to Small Home Screen Widgets. For them to appear nicely, I changed the order of value title and value and made the value font larger. They can be added on the iPad Lock Screen with a Horizontal Orientation.

These are the 2 variations that system small widgets can be presented in:

Small (Lock Screen)
Small (Home Screen)

The inspiration was a weather widget example:

weather 2
Weather Beforre

## Lock Screen Widgets (Accessory Rectangular)

To make the selection consistent with iOS 17 design guidelines and newest widgets, I think the best option is to remove the background altogether:

IMG_1040

Now all these widgets can be displayed together on the iPad Lock Screen:

Simulator Screenshot - iPad Air (5th generation) - 2023-10-18 at 11 34 01

**Comment by staskus** [comment]
@guarani 

My plan is as follows:

1. Get feedback from designers about widgets without background pc8eDl-14z-p2#comment-935
2. Split this PR into two: 1) Widget Lock Screen Widget UI updates without a background and 2) iOS 17 APIs conformance
3. Review, test & merge **PR 1)** since it would not require Xcode changes, mostly it would remove background from lock screen widgets
4. Review, test & merge https://github.com/wordpress-mobile/WordPress-iOS/pull/21514, since it would be counter-productive to test it while this PR is still ongoing
5. Ideally, we would ship lock screen widget background updates and this week widget with 23.6
6. Once we enable Xcode 15, merge **2) iOS 17 APIs conformance** PR.

**Comment by guarani** [comment]
FYI: The PR that updates CI to Xcode 15 is targeting Monday's 24.0 release: https://github.com/wordpress-mobile/WordPress-iOS/pull/21921

We could spend some time this week to get necessary widget changes done for Monday's beta cut. Wdyt? I haven't been focusing on these PRs but I think it makes sense to do so now.

**Comment by staskus** [comment]
> FYI: The PR that updates CI to Xcode 15 is targeting Monday's 24.0 release: https://github.com/wordpress-mobile/WordPress-iOS/pull/21921

@guarani, thanks for making a note! Yes, I followed it but I wasn't sure if it would make it on time for 24.0. 

> Wdyt? I haven't been focusing on these PRs but I think it makes sense to do so now.

Yes, once they merge Xcode 15 updates we'll need to get this PR merged for 24.0 as well. It would be good to build it with Xcode 15 locally and review it so it would be ready. 👍

**Review by mokagio — COMMENTED** [review]
I'll start by saying that I haven't looked at the whole code, so apologies if the question has an easy answer. But I wonder if any of this could go into [`JetpackStatsWidgetsCore`](https://github.com/wordpress-mobile/WordPress-iOS/tree/4e9bb83c82264e642cdd36e4456ff1963577d3a7/Modules/Sources/JetpackStatsWidgetsCore).

**Comment by staskus** [comment]
> I'll start by saying that I haven't looked at the whole code, so apologies if the question has an easy answer. But I wonder if any of this could go into [`JetpackStatsWidgetsCore`](https://github.com/wordpress-mobile/WordPress-iOS/tree/4e9bb83c82264e642cdd36e4456ff1963577d3a7/Modules/Sources/JetpackStatsWidgetsCore).

Great question, @mokagio. I'll check! I made this PR 3 months ago to prepare for Xcode 15 updates when this module didn't exist.

**Comment by staskus** [comment]
> I'll start by saying that I haven't looked at the whole code, so apologies if the question has an easy answer. But I wonder if any of this could go into [JetpackStatsWidgetsCore](https://github.com/wordpress-mobile/WordPress-iOS/tree/4e9bb83c82264e642cdd36e4456ff1963577d3a7/Modules/Sources/JetpackStatsWidgetsCore).

I checked the affected files and I don't think any of them needs to go to `JetpackStatsWidgetCore`. Changes are only to Stats Widget UI code which can continue belonging to `JetpackStatsWidgets`. `JetpackStatsWidgetCore` should contain data types and functionality that needs to be reused between targets.

File: WordPress/JetpackStatsWidgets/Views/SingleStatView.swift
Comment (mokagio) [inline WordPress/JetpackStatsWidgets/Views/SingleStatView.swift:17]: Did you consider delegating to `self.init(title:...`?

File: WordPress/JetpackStatsWidgets/LockScreenWidgets/LockScreenStatsWidget.swift
Comment (mokagio) [inline WordPress/JetpackStatsWidgets/LockScreenWidgets/LockScreenStatsWidget.swift:36]: **Super nitpick**, `///` is a documentation comment, as opposed to the standard `//`. With `///` Xcode and other tools will render the comment as documentation reading tags and markdown. However, the comment needs to be on a type or method declaration. We don't need to "document" method calls. I just wanted to point it out but I don't expect this to be addressed. 👍 😄

File: WordPress/JetpackStatsWidgets/Views/Cards/StatsValueView.swift
Comment (mokagio) [inline WordPress/JetpackStatsWidgets/Views/Cards/StatsValueView.swift:33]: Why 0.5?

**Comment by mokagio** [comment]
@staskus thank you for looking into whether some files could go into the module.

> `JetpackStatsWidgetCore` should contain data types and functionality that needs to be reused between targets.

Yes, and also any kind of standalone logic that we might want (ought to!) unit test. We can't run unit tests on app extension targets, so anything we want to test should go into that Swift Package. 

As far as I can see, this PR is all about SwiftUI configurations so there's nothing that fits that criteria either 👍

File: WordPress/JetpackStatsWidgets/LockScreenWidgets/Views/LockScreenMultiStatView.swift
Comment (mokagio) [inline WordPress/JetpackStatsWidgets/LockScreenWidgets/Views/LockScreenMultiStatView.swift:38]: Just curious about the Swift UI architecture, which is not my strong suit. Would it have been possible to group the conditional logic, e.g. in an `*Stack`, to only write the `removableWidgetBackground()` transformation once?

**Review by mokagio — APPROVED** [review]
Run it on the Simulator with Xcode 15.1 after merging `trunk` into this branch. Everything looks good as far as I can see, but I have little context around the expected widgets look and feel.

Let's merge this so it's part of 24.0, get feedback from @osullivanchris (Chris has already commented on the widgets, see pc8eDl-14z-p2), and address possible tweaks as beta fixes.

### iPhone

<table>
<tr>
<td>
Screenshot 2024-01-08 at 16 37 04
</td>
<td>
image
</td>
<td>
Screenshot 2024-01-08 at 16 43 45
</td>
</tr>
</table>

### iPad

<table>
<tr>
<td>
Screenshot 2024-01-08 at 16 52 40

</td>
<td>

Screenshot 2024-01-08 at 16 53 38


</td>
</tr>
</table>

File: WordPress/JetpackStatsWidgets/LockScreenWidgets/Views/LockScreenMultiStatView.swift
Comment (staskus) [reply WordPress/JetpackStatsWidgets/LockScreenWidgets/Views/LockScreenMultiStatView.swift:38]: Good question. I think it would be possible 👍

**Comment by staskus** [comment]
Thanks for the review, @mokagio!

I can merge it to 24.0. 

This PR was all about basic iOS 17 support. It meant supporting `removableWidgetBackground` and also making sure existing medium and large widgets could appear without a background on the iPad OS Horizontal Lock Screen.

I agree we can make other tweaks after we merge this PR.

File: WordPress/JetpackStatsWidgets/Views/SingleStatView.swift
Comment (staskus) [reply WordPress/JetpackStatsWidgets/Views/SingleStatView.swift:17]: Good observation. I'll use `self.init..`

File: WordPress/JetpackStatsWidgets/Views/Cards/StatsValueView.swift
Comment (staskus) [reply WordPress/JetpackStatsWidgets/Views/Cards/StatsValueView.swift:33]: 0.5 is an arbitrary value, could be smaller or a bit larget as well.

Previously created rectangular widgets can now appear without a background on iPadOS Horizontal Lock Screen. Since they appear without a background, I increased font sizes to have less space and resemble the look of other widgets without a background.

However, when numbers become large, I want font to Value font to shrink, to be fully displayed.

Without `minimumScaleFactor`:
image

With `minimumScaleFactor`:
image

**Comment by mokagio** [comment]
I think the fact the latest build on this branch was cancelled made all the CI checks go "Errored"

image

I started a new one so we can merge this ASAP.

@staskus what do you mean with "I can merge it to 24.0." and with the "[Status] Blocked" label? I read this as "there's a reason why we shouldn't merge it to `trunk` right now" but I don't see what the reason is 🤔

**Comment by staskus** [comment]
> I started a new one so we can merge this ASAP.

@mokagio 

Okay. Sorry, there was a bit of a misunderstanding. I missed the fact you already merged it with trunk. Yes, we can merge it whenever CI succeeds. 👍

**Comment by guarani** [comment]
I forgot to review on Friday 🙈  Thanks for handling this folks!

**Pull request merged**
